### ISSUE TITLE
Text with buggy UTF-8 character in Flutter 3.3.1

### ISSUE THREAD

**Issue by andgar2010** [flutter/flutter#111377]
## Steps to Reproduce

![image](https://user-images.githubusercontent.com/19521054/189551449-6e791aa5-f104-4527-81ae-114b6db1c2f4.png)

![image](https://user-images.githubusercontent.com/19521054/189554247-ffa308f7-13d6-495f-94e4-882566de25b9.png)

![image](https://user-images.githubusercontent.com/19521054/189553455-a83abeba-da3b-438c-b006-1c2919ff54d7.png)

![image](https://user-images.githubusercontent.com/19521054/189551438-0fcff735-4dde-4d80-8bea-9589ee0bf1a1.png)

<details>
<summary>Code sample</summary>

Data JSON:
```json
...
    {
      "alpha_two_code": "US",
      "domains": [
        "nmmi.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.nmmi.edu"
      ],
      "name": "New Mexico Military Institute"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "nmsua.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://nmsua.edu"
      ],
      "name": "New Mexico State University-​Alamogordo"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "cavern.nmsu.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.cavern.nmsu.edu/"
      ],
      "name": "New Mexico State University-​Carlsbad"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "dabcc.nmsu.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://dabcc.nmsu.edu/"
      ],
      "name": "New Mexico State University-​Dona Ana"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "sanjuancollege.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.sanjuancollege.edu"
      ],
      "name": "San Juan College"
    },
   {
      "alpha_two_code": "US",
      "domains": [
        "jacksonville-college.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.jacksonville-college.edu"
      ],
      "name": "Jacksonville College-​Main Campus"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "kilgore.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.kilgore.edu"
      ],
      "name": "Kilgore College"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "lit.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.lit.edu"
      ],
      "name": "Lamar Institute of Technology"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "lsco.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.lsco.edu"
      ],
      "name": "Lamar State College-​Orange"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "lamarpa.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.lamarpa.edu"
      ],
      "name": "Lamar State College-​Port Arthur"
    },
    {
      "alpha_two_code": "US",
      "domains": [
        "laredo.edu"
      ],
      "country": "United States",
      "state-province": null,
      "web_pages": [
        "http://www.laredo.edu"
      ],
      "name": "Laredo Community College"
    },
...
```

```dart
ListView.builder(
  itemCount: universities.length,
  itemBuilder: (BuildContext context, int index) {
    final UniversityModel? item = universities[index];

    return ListTile(
      title: Text(item?.name ?? ''),
      leading: const CircleAvatar(
        foregroundImage:
            AssetImage('assets/images/flutter_logo.png'),
      ),
      onTap: () {
        Navigator.restorablePushNamed(
          context,
          UniversityItemDetailsView.routeName,
        );
      },
    );
  },
)
```
</details>

<details>
  <summary>Logs</summary>

```log
[  +84 ms] --------- beginning of main
           09-11 22:41:55.569 V/MediaPlayerList(  762): mPackageChangedBroadcastReceiver: action: android.intent.action.PACKAGE_REMOVED
[   +5 ms] executing: C:\Users\AndGar\AppData\Local\Android\sdk\platform-tools\adb.exe -s emulator-5554 shell am start -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -f 0x20000000 --ez enable-dart-profiling true --ez enable-checked-mode true --ez verify-entry-points true com.example.demo_university_ag/com.example.demo_university_ag.MainActivity
[ +311 ms] Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x20000000 cmp=com.example.demo_university_ag/.MainActivity (has extras) }
[        ] Waiting for observatory port to be available...
[+2907 ms] Observatory URL on device: http://127.0.0.1:44401/CaVpSxy5m6g=/
[        ] executing: C:\Users\AndGar\AppData\Local\Android\sdk\platform-tools\adb.exe -s emulator-5554 forward tcp:0 tcp:44401
[  +55 ms] 5738
[        ] Forwarded host port 5738 to device port 44401 for Observatory
[   +5 ms] Caching compiled dill
[  +30 ms] Connecting to service protocol: http://127.0.0.1:5738/CaVpSxy5m6g=/
[+1398 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at http://127.0.0.1:5738/CaVpSxy5m6g=/.
[ +343 ms] DDS is listening at http://127.0.0.1:5745/NAYdAwpaX_s=/.
[  +56 ms] Successfully connected to service protocol: http://127.0.0.1:5738/CaVpSxy5m6g=/
[  +96 ms] DevFS: Creating new filesystem on the device (null)
[  +30 ms] DevFS: Created new filesystem on the device (file:///data/user/0/com.example.demo_university_ag/code_cache/demo_university_agYAVENA/demo_university_ag/)
[   +1 ms] Updating assets
[ +218 ms] Manifest contained wildcard assets. Inserting missing file into build graph to force rerun. for more information see #56466.
[   +4 ms] Syncing files to device sdk gphone64 x86 64...
[   +1 ms] Compiling dart to kernel with 0 updated files
[        ] Processing bundle.
[        ] <- recompile package:demo_university_ag/main.dart 2500a54f-0799-4977-b9c2-30bc0d4a3807
[        ] <- 2500a54f-0799-4977-b9c2-30bc0d4a3807
[  +20 ms] Bundle processing done.
[ +130 ms] Updating files.
[        ] DevFS: Sync finished
[        ] Syncing files to device sdk gphone64 x86 64... (completed in 153ms)
[        ] Synced 0.0MB.
[   +1 ms] <- accept
[   +5 ms] Connected to _flutterView/0x784d71754110.
[   +2 ms] Flutter run key commands.
[   +1 ms] r Hot reload.
[        ] R Hot restart.
[        ] h List all available interactive commands.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ]  Running with sound null safety
[        ] An Observatory debugger and profiler on sdk gphone64 x86 64 is available at: http://127.0.0.1:5745/NAYdAwpaX_s=/
[+1137 ms] The Flutter DevTools debugger and profiler on sdk gphone64 x86 64 is available at: http://127.0.0.1:9101?uri=http://127.0.0.1:5745/NAYdAwpaX_s=/
[+1077 ms] D/EGL_emulation( 5093): app_time_stats: avg=57.55ms min=11.37ms max=433.18ms count=18
[+1019 ms] D/EGL_emulation( 5093): app_time_stats: avg=19.59ms min=3.67ms max=77.64ms count=50
[+4986 ms] D/EGL_emulation( 5093): app_time_stats: avg=355.87ms min=12.62ms max=4261.60ms count=14
[+1007 ms] D/EGL_emulation( 5093): app_time_stats: avg=47.89ms min=26.65ms max=95.10ms count=21
[+1006 ms] D/EGL_emulation( 5093): app_time_stats: avg=23.39ms min=7.90ms max=50.98ms count=41
[+1010 ms] D/EGL_emulation( 5093): app_time_stats: avg=37.88ms min=8.34ms max=73.43ms count=26
[+1015 ms] D/EGL_emulation( 5093): app_time_stats: avg=31.39ms min=10.75ms max=53.58ms count=32
[+1011 ms] D/EGL_emulation( 5093): app_time_stats: avg=30.15ms min=8.50ms max=48.82ms count=33
[+1036 ms] D/EGL_emulation( 5093): app_time_stats: avg=24.58ms min=9.03ms max=90.75ms count=41
[+1020 ms] D/EGL_emulation( 5093): app_time_stats: avg=30.30ms min=7.09ms max=78.44ms count=33
[+1023 ms] D/EGL_emulation( 5093): app_time_stats: avg=32.51ms min=6.92ms max=73.04ms count=31
[+1040 ms] D/EGL_emulation( 5093): app_time_stats: avg=31.69ms min=9.41ms max=82.50ms count=32
[+1014 ms] D/EGL_emulation( 5093): app_time_stats: avg=25.17ms min=10.59ms max=55.37ms count=39
[+1006 ms] D/EGL_emulation( 5093): app_time_stats: avg=25.23ms min=7.34ms max=55.77ms count=39
[+1014 ms] D/EGL_emulation( 5093): app_time_stats: avg=36.01ms min=13.93ms max=87.84ms count=28
[+1006 ms] D/EGL_emulation( 5093): app_time_stats: avg=20.67ms min=8.13ms max=47.84ms count=46
[+1023 ms] D/EGL_emulation( 5093): app_time_stats: avg=28.30ms min=13.84ms max=46.94ms count=36
[+1029 ms] D/EGL_emulation( 5093): app_time_stats: avg=34.28ms min=15.60ms max=65.48ms count=30
[+1003 ms] D/EGL_emulation( 5093): app_time_stats: avg=29.11ms min=7.47ms max=61.88ms count=34
[+1009 ms] D/EGL_emulation( 5093): app_time_stats: avg=16.54ms min=5.69ms max=40.54ms count=58
[+1008 ms] D/EGL_emulation( 5093): app_time_stats: avg=25.59ms min=6.70ms max=54.09ms count=38
[+1016 ms] D/EGL_emulation( 5093): app_time_stats: avg=21.92ms min=6.80ms max=65.90ms count=45
[+1037 ms] D/EGL_emulation( 5093): app_time_stats: avg=30.84ms min=8.16ms max=75.31ms count=33
[+1011 ms] D/EGL_emulation( 5093): app_time_stats: avg=20.82ms min=8.23ms max=93.61ms count=46
[+1007 ms] D/EGL_emulation( 5093): app_time_stats: avg=17.97ms min=5.71ms max=73.17ms count=54
[+1003 ms] D/EGL_emulation( 5093): app_time_stats: avg=20.66ms min=5.07ms max=186.59ms count=47
[+1004 ms] D/EGL_emulation( 5093): app_time_stats: avg=21.07ms min=6.77ms max=100.85ms count=47
[+1005 ms] D/EGL_emulation( 5093): app_time_stats: avg=18.30ms min=6.70ms max=35.04ms count=53
[+1379 ms] D/EGL_emulation( 5093): app_time_stats: avg=89.54ms min=5.78ms max=989.50ms count=15
[+1010 ms] D/EGL_emulation( 5093): app_time_stats: avg=17.00ms min=5.18ms max=34.18ms count=55
[+1023 ms] D/EGL_emulation( 5093): app_time_stats: avg=18.99ms min=4.88ms max=90.58ms count=48
[+1345 ms] D/EGL_emulation( 5093): app_time_stats: avg=30.60ms min=7.00ms max=562.81ms count=43
[+1004 ms] D/EGL_emulation( 5093): app_time_stats: avg=19.56ms min=6.86ms max=142.64ms count=48
[+1009 ms] D/EGL_emulation( 5093): app_time_stats: avg=20.47ms min=7.90ms max=168.21ms count=47
[+1000 ms] D/EGL_emulation( 5093): app_time_stats: avg=19.78ms min=4.01ms max=82.85ms count=48
[+1016 ms] D/EGL_emulation( 5093): app_time_stats: avg=17.67ms min=7.00ms max=93.53ms count=56
[+1114 ms] D/EGL_emulation( 5093): app_time_stats: avg=30.18ms min=8.46ms max=405.49ms count=36
[+1215 ms] D/EGL_emulation( 5093): app_time_stats: avg=73.55ms min=7.24ms max=485.97ms count=16
```

![image](https://user-images.githubusercontent.com/19521054/189551820-125e488c-c4cc-401d-99a3-edcb1b678c99.png)

```
$ flutter doctor -v
[√] Flutter (Channel stable, 3.3.1, on Microsoft Windows [Versi¢n 10.0.19044.1949], locale es-CO)
    • Flutter version 3.3.1 on channel stable at C:\Users\AG\fvm\versions\3.3.1
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4f9d92fbbd (5 days ago), 2022-09-06 17:54:53 -0700
    • Engine revision 3efdf03e73
    • Dart version 2.18.0
    • DevTools version 2.15.0

[√] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at C:\Users\AndGar\AppData\Local\Android\sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.2.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.2.32519.379
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code, 64-bit edition (version 1.71.0)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.48.0

[√] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 12 (API 32) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Versi¢n 10.0.19044.1949]
    • Chrome (web)                 • chrome        • web-javascript • unknown
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 105.0.1343.33

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by exaby73** [comment]
Hello @andgar2010. Thank you for filing this issue. Can you please provide the following information?

- A [complete and minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Ideally, a single `main.dart` file can can be run directly.

**Comment by FusoraTech** [comment]
As we can see observing the provided code, the string contains a hidden UTF-8 char called ZWSP (Zero width space, \u{200b} utf8 bytes e2 80 8b ) and Flutter is rendering it wrongly.

![image](https://user-images.githubusercontent.com/95307527/189891311-ae55883c-987b-4dae-8f87-c649413bbfb5.png)

The ZWSP character is intended for invisible word separation and for line break control; it has no width, but its presence between two characters does not prevent increased letter spacing in justification

**Comment by exaby73** [comment]
Can you provide sample code that would allow me to verify the issue?

**Comment by andgar2010** [comment]
Hello @exaby73. Thank you for you response. Sorry later response. I send [a reproducible example]( https://dartpad.dev/?id=f438f31507b42b1968d68fb5a0f90574) on DartPad

**Comment by exaby73** [comment]
The URL you included in the Dartpad is being blocked by CORS. Could you provide a sample with the needed text in the code sample itself? If the data returned on the URL ever changes or the URL expires, this issue would become unactionable.

**Comment by andgar2010** [comment]
Hi, sorry. Already corrected for the data in gist. I did the test in browser presents works without problems, however, the same code in android presents a replication error.

**Comment by exaby73** [comment]
With the updated gist's code, I do not see any problems in UI. Tested on Windows, Web and Android.

<details>
<summary>Screenshot</summary>

<img width="300" src="https://user-images.githubusercontent.com/50555895/190331316-1156164d-f930-405a-b860-8cefa89c390e.png">

</details>

### Flutter Doctor

<details><summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.3.2, on Microsoft Windows [Version 10.0.22000.978], locale en-US)
    • Flutter version 3.3.2 on channel stable at C:\Users\Nabeel\fvm\versions\stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (17 hours ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.71.0)
    • VS Code at C:\Users\Nabeel\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.978]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 105.0.1343.27

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by andgar2010** [comment]
I have checked it with the v3.3.2 update that comes out working on Android and I can't find any bugs. it seems to be fixed with v3.3.2. this case be closed.
